### ISSUE TITLE
CORE Spec - Org Management Component - Identity Management  - High level diagrams

### ISSUE THREAD

**Issue by Consianimis** [ethereum-oasis/baseline-standard#28]
**Purpose:**

The purpose is to provide non-normative examples of different identity management architectures that the Baseline Protocol specification Must/Should support. Assumptions for each architecture should be documented and high-level diagrams should be provided.
- federated identity management architecture
- decentralized identity management architecture
- other proposed ?

**Comment by awoie** [comment]
Before we can come up with a high-level diagram, let's clarify some of the assumptions regarding the registration of a legal entity (no baselining yet with counterparties) ...

Actors DID/VC approach:
- DID registrar that creates the DID for a legal entity which can be either a service (e.g., Azure services for did:ion) or just a local library. In general, this is out of scope of the baseline protocol. It is just assumed the DID will be created outside of IRegistry.
- DIDs will have messaging endpoints as well as public cryptographic key material used for baseline. 
- VC registrar which can be either a service (e.g., W3C VC HTTP API, e.g., Azure services) or just some library. Same as DID registrar and out of scope of baseline. We might later specify a data model (aka credential schema/context) for VCs at least.
- Legal entity interacting with the registrars above and the IRegistry implementation.
- IRegistry implementation supports DIDs and VCs. The registry will have some policies configured that allows the registry to trust certain DIDs and/or VCs, e.g., trusted issuers, domain linkage VC.
- Employees of legal entities. They will not be a dedicated actor as they will act on behalf of the legal entity. I guess, authentication of the employees is out of scope as well?

Actors Federated approach:
- Assuming we focus on OpenID Connect and I haven't heard any requirements from the community so far except that this should be possible in general.
- Identity Provider which is an OpenID Connect Provider (OP) that is used by a working group. Out of scope of the baseline protocol but it is assumed that it provides identity assertions, i.e., id_tokens.
- IRegistry implementation that supports verification of OpenID Connect tokens (id_token) and has a trust relationship with the OP issuer (e.g., through pinning the public key JWK URI). Trust configuration is out of scope of the IRegistry interface but in general we can assume there will be a policy that verifies an id_token. The id_token would also contain the IRegistry as the audience (`aud`). We can also discuss whether the IRegistry itself is the OIDC RP, any thoughts?
- Legal entity that has a client_id with the OP. Legal entities act as Relying Parties (RPs) in OIDC.
- Employees have user accounts at the OP and can login to the RP. They would also act on behalf of the legal entity but they would like have their own identifier in the id_token.

For both options we likely need a common data model for the identity data that can be used in the IRegistry API. To come up with a suggestion, we would need to clarify the questions above.

@Therecanbeonlyone1969 @caleteeter @kthomas ?

**Comment by Therecanbeonlyone1969** [comment]
> Before we can come up with a high-level diagram, let's clarify some of the assumptions regarding the registration of a legal entity (no baselining yet with counterparties) ...
> 

Yes

> Actors DID/VC approach:
> 
> * DID registrar that creates the DID for a legal entity which can be either a service (e.g., Azure services for did:ion) or just a local library. In general, this is out of scope of the baseline protocol. It is just assumed the DID will be created outside of IRegistry.

AF: Yes

> * DIDs will have messaging endpoints as well as public cryptographic key material used for baseline.

AF: Yes -- there exists one or more endpoints that allow a communication endpoint to be accessed. Note, that the communication endpoint does not necessarily have to be in the DIDdoc.

> * VC registrar which can be either a service (e.g., W3C VC HTTP API, e.g., Azure services) or just some library. Same as DID registrar and out of scope of baseline. We might later specify a data model (aka credential schema/context) for VCs at least.

AF: Yes

> * Legal entity interacting with the registrars above and the IRegistry implementation.

AF: Yes

> * IRegistry implementation supports DIDs and VCs. The registry will have some policies configured that allows the registry to trust certain DIDs and/or VCs, e.g., trusted issuers, domain linkage VC.

AF: Yes

> * Employees of legal entities. They will not be a dedicated actor as they will act on behalf of the legal entity. I guess, authentication of the employees is out of scope as well?

AF: Yes

> Actors Federated approach:
> 
> * Assuming we focus on OpenID Connect and I haven't heard any requirements from the community so far except that this should be possible in general.

AF: Yes, for simplicity

> * Identity Provider which is an OpenID Connect Provider (OP) that is used by a working group. Out of scope of the baseline protocol but it is assumed that it provides identity assertions, i.e., id_tokens.

AF: Yes

> * IRegistry implementation that supports verification of OpenID Connect tokens (id_token) and has a trust relationship with the OP issuer (e.g., through pinning the public key JWK URI). Trust configuration is out of scope of the IRegistry interface but in general we can assume there will be a policy that verifies an id_token. The id_token would also contain the IRegistry as the audience (`aud`). We can also discuss whether the IRegistry itself is the OIDC RP, any thoughts?

AF: Yes. However, we could think about defining a security profile as part of the standard!

> * Legal entity that has a client_id with the OP. Legal entities act as Relying Parties (RPs) in OIDC.

AF: Yes

> * Employees have user accounts at the OP and can login to the RP. They would also act on behalf of the legal entity but they would like have their own identifier in the id_token.

AF: Yes. Normal OIDC model.

> For both options we likely need a common data model for the identity data that can be used in the IRegistry API. To come up with a suggestion, we would need to clarify the questions above.
> 
> @Therecanbeonlyone1969 @caleteeter @kthomas ?

@awoie @chaals @kthomas

**Comment by awoie** [comment]
@Therecanbeonlyone1969 As a starting point, I created the attached diagram. Note, in that diagram the operator of the registry itself is not the operator of the company. 

![basline-overview](https://user-images.githubusercontent.com/43441584/112317438-5861cc80-8cac-11eb-88e7-e35baa7de89d.png)

**Comment by awoie** [comment]
- The identity system can either be OP-based or Decentralized Identity-based. 
- The IRegistry implementation can either be an OIDC RP or Decentralized Identity verifier.

![baseline-id2](https://user-images.githubusercontent.com/43441584/112321440-48e48280-8cb0-11eb-9a3d-ee2a809a241a.png)

**Comment by awoie** [comment]
There is one things I cannot get my head around with...

Assume you are using the DID approach.

Why would you have to notify the Registry about updates to message endpoints and public cryptographic material at all? Meaning, why would you need go through the registry's API if the information can be derived directly from the DID. You would risk to run out of sync with the registry data, right?

**Comment by awoie** [comment]
Further, with regard to the federated model / OIDC, there are essentially two approaches:

1. Registry is a registered OIDC RP (client). In that case, the registry would receive the `id_token` directly from the OP according to the OIDC flow. Note, usually, code flow is used and the `id_token` obtained directly in the RP (Registry). That does not fit well with the model described above unless we would rather use OAuth2 (note, OIDC is OAuth2 + identity data) and have no `id_token` at all. The identity data will be provided through the IRegistry API that is protected by an OAuth2 `access_token`.

2. We could support OIDC, if the registry is another `aud` in the `id_token`. In that case, the operator would be the OIDC RP (client) but requesting that the `id_token` will be presented to the Registry. The `id_token` would act as a Bearer token to get access to the IRegistry API. The Registry would need to verify that the `id_token` contains the Registry in the `aud` property in addition.

The challenge here is to find a homogenous approach that works for both models. Any thoughts?

**Comment by Therecanbeonlyone1969** [comment]
> There is one things I cannot get my head around with...
> 
> Assume you are using the DID approach.
> 
> Why would you have to notify the Registry about updates to message endpoints and public cryptographic material at all? Meaning, why would you need go through the registry's API if the information can be derived directly from the DID. You would risk to run out of sync with the registry data, right?

@awoie ... super work! and no you do not have to keep the endpoints in the registry ... only the DID -> Baseline Account KV-pair.

**Comment by Therecanbeonlyone1969** [comment]
@awoie 

> Further, with regard to the federated model / OIDC, there are essentially two approaches:
> 
> 1. Registry is a registered OIDC RP (client). In that case, the registry would receive the `id_token` directly from the OP according to the OIDC flow. Note, usually, code flow is used and the `id_token` obtained directly in the RP (Registry). That does not fit well with the model described above unless we would rather use OAuth2 (note, OIDC is OAuth2 + identity data) and have no `id_token` at all. The identity data will be provided through the IRegistry API that is protected by an OAuth2 `access_token`.

I agree with this. It should be the recommended approach!

> 2. We could support OIDC, if the registry is another `aud` in the `id_token`. In that case, the operator would be the OIDC RP (client) but requesting that the `id_token` will be presented to the Registry. The `id_token` would act as a Bearer token to get access to the IRegistry API. The Registry would need to verify that the `id_token` contains the Registry in the `aud` property in addition.
> 

We should leave this as an option, but not the recommended approach. Keep it simple. Using OAuth for authentication is easy and you can even use a DID within OAuth, if so desired.

> The challenge here is to find a homogenous approach that works for both models. Any thoughts?

We need to separate out three things IMHO:
1. How did Company A establish a trust relationship with a Baseline Operator aka there is an account that will be created by the Basline Operator for Company A somehow that contains a "secret" to be used to get an OAuth Bearer Token from the Baseline Operator for endpoint authentication? -> here you have an Idp either an OP or a VC issuer e.g. MSFT for Azure AD account holders
2. Authentication for endpoints (DIDauth or OAuth)
3. Authorization (OAuth/OIDC or VCs)

Makes sense?
